pkg/controller/utils/pod/pod_test.go

```diff
@@ -18,7 +18,354 @@ import (
+					Terminated: &v1.ContainerStateTerminated{
+						Reason:     "CrashLoopBackOff",
+						FinishedAt: metav1.NewTime(now.Add(-time.Hour)),
+					},
+				},
+			},
+			{
+				RestartCount: 1,
+				LastTerminationState: v1.ContainerState{
+					Terminated: &v1.ContainerStateTerminated{
+						Reason:     "OOMKilled",
+						FinishedAt: metav1.NewTime(now.Add(-2 * time.Hour)),
+					},
+				},
+			},
+		},
+	},
+	)
+
+	want := "node1"
+	got, isScheduled := IsPodScheduled(pod)
+	assert.Equal(t, want, got)
+	assert.True(t, isScheduled)
+
+	pod2 := ctrltest.NewPod("bar", "pod2", "", &ctrltest.NewPodOptions{
+		ContainerStatuses: []v1.ContainerStatus{
+			{
+				RestartCount: 10,
+				LastTerminationState: v1.ContainerState{
+					Terminated: &v1.ContainerStateTerminated{
+						Reason:     "CrashLoopBackOff",
+						FinishedAt: metav1.NewTime(now.Add(-time.Hour)),
+					},
+				},
+			},
+			{
+				RestartCount: 1,
+				LastTerminationState: v1.ContainerState{
+					Terminated: &v1.ContainerStateTerminated{
+						Reason:     "OOMKilled",
+						FinishedAt: metav1.NewTime(now.Add(-2 * time.Hour)),
+					},
+				},
+			},
+		},
+	},
+	)
+	got, isScheduled = IsPodScheduled(pod2)
+	assert.Equal(t, "", got)
+	assert.False(t, isScheduled)
+}
+
+func TestGetNodeNameFromPod(t *testing.T) {
+	now := time.Now()
+	pod := ctrltest.NewPod("bar", "pod1", "node1", &ctrltest.NewPodOptions{
+		ContainerStatuses: []v1.ContainerStatus{
+			{
+				RestartCount: 10,
+				LastTerminationState: v1.ContainerState{
+					Terminated: &v1.ContainerStateTerminated{
+						Reason:     "CrashLoopBackOff",
+						FinishedAt: metav1.NewTime(now.Add(-time.Hour)),
+					},
+				},
+			},
+			{
+				RestartCount: 1,
+				LastTerminationState: v1.ContainerState{
+					Terminated: &v1.ContainerStateTerminated{
+						Reason:     "OOMKilled",
+						FinishedAt: metav1.NewTime(now.Add(-2 * time.Hour)),
+					},
+				},
+			},
+		},
+	},
+	)
+	want := "node1"
+	got, err := GetNodeNameFromPod(pod)
+	assert.Equal(t, want, got)
+	assert.Nil(t, err)
+
+	pod2 := ctrltest.NewPod("bar", "pod2", "", &ctrltest.NewPodOptions{
+		ContainerStatuses: []v1.ContainerStatus{
+			{
+				RestartCount: 1,
+				LastTerminationState: v1.ContainerState{
+					Terminated: &v1.ContainerStateTerminated{
+						Reason:     "OOMKilled",
+						FinishedAt: metav1.NewTime(now.Add(-2 * time.Hour)),
+					},
+				},
+			},
+		},
+	},
+	)
+	got, err = GetNodeNameFromPod(pod2)
+	assert.Equal(t, "", got)
+	assert.NotNil(t, err)
+}
+
+func TestIsPodReady(t *testing.T) {
+	pod := ctrltest.NewPod("bar", "pod1", "node1", &ctrltest.NewPodOptions{})
+	isReady := IsPodReady(pod)
+	assert.False(t, isReady)
+
+	pod2 := ctrltest.NewPod("bar", "pod2", "node1", &ctrltest.NewPodOptions{})
+	pod2.Status.Conditions = []v1.PodCondition{
+		{
+			Type:   v1.PodReady,
+			Status: v1.ConditionTrue,
+		},
+	}
+	isReady = IsPodReady(pod2)
+	assert.True(t, isReady)
+}
+
+func TestIsCannotStartReason(t *testing.T) {
+	reason := "ErrImagePull"
```

Comment:
Is it worth to check all the reasons? Something like (not tested):
```
allReasons := _get_from_code_or_define_here_
for _, reason := range allReasons {
    cannotStart := IsCannotStartReason(reason)
    assert.True(t, cannotStart)
}
```

Reply:
i like that, thanks